registry/storage/tagstore_test.go

```diff
@@ -2,29 +2,78 @@ package storage
 
 import (
 	"context"
+	"errors"
+	"io"
 	"reflect"
+	"strconv"
 	"testing"
 
 	"github.com/distribution/distribution/v3"
 	"github.com/distribution/distribution/v3/manifest"
 	"github.com/distribution/distribution/v3/manifest/schema2"
 	"github.com/distribution/distribution/v3/reference"
+	storagedriver "github.com/distribution/distribution/v3/registry/storage/driver"
+	"github.com/distribution/distribution/v3/registry/storage/driver/base"
 	"github.com/distribution/distribution/v3/registry/storage/driver/inmemory"
 	digest "github.com/opencontainers/go-digest"
 )
 
 type tagsTestEnv struct {
-	ts  distribution.TagService
-	bs  distribution.BlobStore
-	ms  distribution.ManifestService
-	gbs distribution.BlobStatter
-	ctx context.Context
+	ts         distribution.TagService
+	bs         distribution.BlobStore
+	ms         distribution.ManifestService
+	gbs        distribution.BlobStatter
+	ctx        context.Context
+	mockDriver *mockInMemory
+}
+
+type mockInMemory struct {
+	base.Base
+	driver          *inmemory.Driver
+	GetContentError error
+}
+
+var _ storagedriver.StorageDriver = &mockInMemory{}
+
+func (m *mockInMemory) Name() string {
+	return m.driver.Name()
+}
+func (m *mockInMemory) GetContent(ctx context.Context, path string) ([]byte, error) {
+	if m.GetContentError != nil {
+		return nil, m.GetContentError
+	}
+	return m.driver.GetContent(ctx, path)
+}
+func (m *mockInMemory) PutContent(ctx context.Context, path string, content []byte) error {
+	return m.driver.PutContent(ctx, path, content)
+}
+func (m *mockInMemory) Reader(ctx context.Context, path string, offset int64) (io.ReadCloser, error) {
+	return m.driver.Reader(ctx, path, offset)
+}
+func (m *mockInMemory) Writer(ctx context.Context, path string, append bool) (storagedriver.FileWriter, error) {
+	return m.driver.Writer(ctx, path, append)
+}
+func (m *mockInMemory) List(ctx context.Context, path string) ([]string, error) {
+	return m.driver.List(ctx, path)
+}
+func (m *mockInMemory) Move(ctx context.Context, sourcePath string, destPath string) error {
+	return m.driver.Move(ctx, sourcePath, destPath)
+}
+func (m *mockInMemory) Delete(ctx context.Context, path string) error {
+	return m.driver.Delete(ctx, path)
+}
+func (m *mockInMemory) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
+	return m.driver.URLFor(ctx, path, options)
+}
+func (m *mockInMemory) Walk(ctx context.Context, path string, f storagedriver.WalkFn) error {
+	return m.driver.Walk(ctx, path, f)
 }
```

Comment:
Struct embedding can get rid of most of the boilerplate.

```suggestion
type mockInMemory struct {
	*inmemory.Driver
	GetContentError error
}

var _ storagedriver.StorageDriver = &mockInMemory{}

func (m *mockInMemory) GetContent(ctx context.Context, path string) ([]byte, error) {
	if m.GetContentError != nil {
		return nil, m.GetContentError
	}
	return m.Driver.GetContent(ctx, path)
}
```